transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -549,6 +554,11 @@ protected void cleanup() {
                 logger.warn("Failed to close the epoll fd.", e);
             }
             try {
+                // Ensure any inflight wakeup writes have been performed prior to closing eventFd.
+                // Cap number of iterations to avoid indefinite spin in the case of a failed write.
+                for (int i = 0; eventFdWriteCount != 0L && i < 10000; i++) {
+                    eventFdWriteCount -= Native.eventFdRead(eventFd.intValue());
+                }
```

Comment:
I thought spinning on read here would be simpler/safer than using epoll, it should rarely ever actually happen. Polling is still an option if others think that's preferable, though we would want probably want to use timerfd as an equivalent safeguard.

Reply:
Btw by "polling" in the comment above I meant blocking wait using `epoll_wait` ... sorry if that was misleading.

Reply:
I was considering calling epollWait here (with a timeout, sure; and we'd need to do epollFd.close() after this). That by itself would be broken, because `epollWaitNow()` is called from `closeAll()`. But I can't figure out why that call to epoll is even there. The best idea I have is that it became vestigial via https://github.com/netty/netty/commit/d63c9f2bdb82c7d686a29a865d37ab3a5455d488 ; previously the `ready` return value was used.

Reply:
Oh yeah I'd wondered about the purpose of that `epollWaitNow` in `closeAll`, hard to imagine it has one now... hopefully @normanmaurer could confirm your suspicion that it's obsolete.

You seemed to imply that `epollWait` by itself would be sufficient (apologies if misinterpreted). But I think the write/read accounting is still needed, i.e. we would still be calling `eventFdRead` in a loop here, just that it would be doing a blocking-wait each iteration rather than spinning.

This is becuase `wakenUp` by itself can't tell us whether a write is still outstanding. Even if we kept track of whether eventfd was in the event array populated by the most recent `epollWait` call, it's possible that there could be more than one outstanding write (from multiple prior EL iterations) and so receiving one last edge wouldn't be conclusive. Do you agree?

But while trying to think about what you may have been proposing I realized that if careful we _could_ in fact do this just by tracking the edges, and so avoid having to read the eventfd after all. Maybe that's what you had in mind and again apologies if so! The key thing is to make sure we don't reset `wakenUp` before waiting during normal loop operation in the case we know a wakeup write is pending... that way we know there can't be more than 1 outstanding at a time.

I have made the corresponding changes here e3305bdcfb5a5673c8a0a32a6d754e11b912d955, and actually think it's a superior option. Let me know what you think and I can update this PR accordingly or open another.

Reply:
> You seemed to imply that epollWait by itself would be sufficient (apologies if misinterpreted). But I think the write/read accounting is still needed

Yes, that's what I was thinking. And yes, just you saying this made me realize that wakeups could overlap in a pathological case. Crap.

> Even if we kept track of whether eventfd was in the event array populated by the most recent epollWait call, it's possible that there could be more than one outstanding write (from multiple prior EL iterations) and so receiving one last edge wouldn't be conclusive. Do you agree?

Yes, I agree.

> But while trying to think about what you may have been proposing I realized that if careful we could in fact do this just by tracking the edges, and so avoid having to read the eventfd after all.

Tell me more! <continues reading>

> The key thing is to make sure we don't reset wakenUp before waiting during normal loop operation in the case we know a wakeup write is pending... that way we know there can't be more than 1 outstanding at a time.

Beautiful! I will say I was looking hard at `wakenUp`, but I hadn't considered anything like this. Honestly, that may make the code easier for us overall, since that sounds less likely to have hidden wakeup races.

Reply:
I looked over e3305bd .  I like it. There is some complexity, but it appears to be a performance optimization (which is fine to me; it just means it isn't as inherently complicated). And we may want to sprinkle a log (even if debug) when epollWaitTimeboxed() times out. And I have a nit for you about reusing the `strategy` int. But I like it. I really like it.

Thank you for noticing the flaw in my logic and then thinking of the solution I should have been thinking of :smile:. (Although I will say, it is a natural extension to the approach; instead of "wait for the notification on shutdown" it is "wait for the notification every time".)

ARE THE HEROS GOING TO BE DONE WITH EPOLL LOOP RACES?
FIND OUT NEXT TIME.

Reply:
Lol, thanks @ejona86, will wait to see what others think and then can hopefully switch over to that and address other comments. Curious which aspect in particular you're referring to re the complexity/optimization. Agree about the logging.

Have a good weekend!

Reply:
@njhill I think the `epollWaitNow()` in closeAll can just be removed...  Most likely it was just ported as we also do a selectNow() in the NIO transport.

Reply:
@normanmaurer let me know if you agree with opening another PR based on e3305bd to replace this one per the above discussion

Reply:
@njhill sounds good.

Reply:
@ejona86 @normanmaurer have now opened #9535 which is just e3305bd rebased.